compiler/semmagic.nim

```diff
@@ -575,7 +575,27 @@ proc magicsAfterOverloadResolution(c: PContext, n: PNode,
     result = n
   of mPrivateAccess:
     var t = n[1].typ[0]
-    if t.kind in {tyRef, tyPtr}: t = t[0]
+    #[
+    find the underlying `object`, even in cases like these:
+    type
+      B[T] = object f0: int
+      A1[T] = ref B[T]
+      A2[T] = ref object f1: int
+      A3 = ref object f2: int
+      A4 = object f3: int
+    ]#
+    var foundPtrLike = false
+    while true:
+      case t.kind
+      of tyRef, tyPtr:
+        if foundPtrLike: break
+        foundPtrLike = true
+        t = t[0]
+      of tyGenericBody: t = t.lastSon
+      of tyGenericInst: t = t[0]
+      of tyGenericInvocation: t = t[0]
```

Comment:
`tyAlias` is missing I think.

And while you're at it, move all this logic to a `semPrivateAccess` proc.

Reply:
done (and added test)